swarm_dial.go

```diff
@@ -529,3 +593,48 @@ func (s *Swarm) dialAddr(ctx context.Context, p peer.ID, addr ma.Multiaddr) (tra
 	// success! we got one!
 	return connC, nil
 }
+
+// TODO We should have a `IsFdConsuming() bool` method on the `Transport` interface in go-libp2p-core/transport.
+// This function checks if any of the transport protocols in the address requires a file descriptor.
+func (s *Swarm) IsFdConsumingAddr(addr ma.Multiaddr) bool {
+	noProxyAddrFnc := func(addr ma.Multiaddr) bool {
+		tpt := s.TransportForDialing(addr)
+		if tpt == nil {
+			// let's return true to be safe
+			log.Debugf("no transport registered for address %s, so marking it as Fd consuming for safety", addr.String())
+			return true
+		}
+		protos := tpt.Protocols()
+		// is a fd consuming transport present in the list of protocols that the transport supports ?
+		for i := range protos {
+			if _, ok := fdConsumingTptProtos[protos[i]]; ok {
+				return true
+			}
+		}
+		return false
+	}
+
+	tpt := s.TransportForDialing(addr)
+	if tpt == nil {
+		// let's return true to be safe
+		log.Debugf("no transport registered for address %s, so marking it as Fd consuming for safety", addr.String())
+		return true
+	}
+
+	if tpt.Proxy() {
+		// check the address of the relay server we will be "directly" dialling.
+		relayaddr, _ := ma.SplitFunc(addr, func(c ma.Component) bool {
+			return c.Protocol().Code == ma.P_CIRCUIT
+		})
+		if relayaddr == nil {
+			log.Debugf("proxy address %s did not have a Circuit Component,"+
+				"so marking it as FD Consuming for safety", addr.String())
+			return true
+		}
+		return noProxyAddrFnc(relayaddr)
+	}
```

Comment:
This isn't necessary. The relay transport doesn't consume a file descriptor.

When the relay transport dials the relay, it'll dial that relay with the swarm. When it does, the swarm will take any fd tokens necessary based on the transport.

Reply:
@Stebalien 

I completely understand that. That's why we have the `dialLimiter.shouldConsumeFd` function which does NOT consume a FD for relays.

However, for partitioning the addresses into FD/Non-FD addresses so we can wait on all dials for the latter before dialing the former, I'd like to **classify** relays where the relay server has a FD consuming addresses as FD consuming so we dial them in the "second batch". This function serves that purpose.